tools/foreman-cockpit-session

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env ruby
+
+require "json"
+require "net/http"
+require "net/ssh"
+
+def dbg(str)
+  # $stderr.write(str + "\n")
+end
+
+# Cockpit protocol, encoding and decoding of control messages.
+
+def send_control(msg)
+  text = JSON.dump(msg)
+  dbg("-> #{text}")
+  $stdout.write("#{text.length+1}\n\n#{text}")
+  $stdout.flush
+end
+
+def read_size(io)
+  size = 0
+  seen = 0
+
+  while seen < 8
```

Comment:
Is there a reason for reading it byte by byte? If not it could probably be done like this

```ruby
def read_size(io)
  data = io.readline.chomp
  raise ArgumentError, 'Invalid frame: size too long' if data.length > 8
  size = data.to_i
  raise ArgumentError, 'Invalid frame: invalid size' if size == 0 && data != '0'
  size
rescue EOFError
  0
end
```

Reply:
Yes, fixed like you propose. We can treat a frame size of zero as an error for control frames, which further simplifies this.